heirloom/cpio/pax.1
'\" t
.\" Copyright (c) 2004 Gunnar Ritter
.\"
.\" This software is provided 'as-is', without any express or implied
.\" warranty. In no event will the authors be held liable for any damages
.\" arising from the use of this software.
.\"
.\" Permission is granted to anyone to use this software for any purpose,
.\" including commercial applications, and to alter it and redistribute
.\" it freely, subject to the following restrictions:
.\"
.\" 1. The origin of this software must not be misrepresented; you must not
.\"    claim that you wrote the original software. If you use this software
.\"    in a product, an acknowledgment in the product documentation would be
.\"    appreciated but is not required.
.\"
.\" 2. Altered source versions must be plainly marked as such, and must not be
.\"    misrepresented as being the original software.
.\"
.\" 3. This notice may not be removed or altered from any source distribution.
.\" Sccsid @(#)pax.1	1.23 (gritter) 4/16/04
.TH PAX 1 "4/16/04" "Heirloom Toolchest" "User Commands"
.SH NAME
pax \- portable archive interchange
.SH SYNOPSIS
.PD 0
.HP
.nh
.ad l
\fBpax\fR [\fB\-cdnvK\fR] [\fB\-b\ \fIsize\fR]
[\fB\-f\ \fIfile\fR] [\fB\-s\ \fIreplstr\fR]
[\fB\-x\ \fIhdr\fR] [\fIpatterns\fR]
.HP
.ad l
\fBpax\fR \fB\-r\fR[\fBcdiknuvK\fR] [\fB\-b\ \fIsize\fR]
[\fB\-f\ \fIfile\fR]
[\fB\-o\ \fIoptions\fR]
[\fB\-p\ \fIpriv\fR] [\fB\-s\ \fIreplstr\fR]
[\fB\-x\ \fIhdr\fR] [\fIpatterns\fR]
.HP
.ad l
\fBpax\fR \fB\-w\fR[\fBadituvLX\fR] [\fB\-b\ \fIsize\fR]
[\fB\-f\ \fIfile\fR]
[\fB\-o\ \fIoptions\fR]
[\fB\-s\ \fIreplstr\fR]
[\fB\-x\ \fIhdr\fR] [\fIfiles\fR]
.HP
.ad l
\fBpax\fR \fB\-rw\fR[\fBdiklntuvLX\fR]
[\fB\-p\ \fIpriv\fR] [\fB\-s\ \fIreplstr\fR]
[\fIfiles\fR] \fIdirectory\fR
.br
.ad b
.hy 1
.PD
.SH DESCRIPTION
.I Pax
creates and extracts file archives and copies files.
.PP
If neither the
.I \-r
or
.I \-w
options are given,
.I pax
works in
.I list
mode
and prints the contents of the archive.
.PP
With the
.B \-r
option,
.I pax
works in
.RI ` read '
mode and extracts files from a file archive.
By default,
the archive is read from standard input.
Optional arguments are interpreted as
.I patterns
and restrict the set of extracted files
to those matching any of the
.IR patterns .
The syntax is identical to that described in
.IR glob (7),
except that the slash character
.RB ` / '
is matched by
meta-character constructs with
.RB ` * ',
.RB ` ? '
and
.RB ` [ '.
Care must be taken to quote meta-characters appropriately from the shell.
If a pattern matches the prefix name of a directory in the archive,
all files below that directory are also extracted.
File permissions are set to those in the archive;
if the caller is the super-user,
ownerships are restored as well.
options are specified.
Archives compressed with
.IR bzip2 (1),
.IR compress (1),
.IR gzip (1),
or
.IR rpm (1)
are transparently de\%compressed on input.
.PP
With
.BR \-w ,
.I pax
works in
.RI ` write '
mode,
creates archives
and writes them to standard output per default.
A list of filenames to be included in the archive is
read from standard input;
if the name of a directory appears,
all its members and the directory itself are recursively
included in the archive.
The
.IR find (1)
utility is useful to generate a list of files
(see also its
.I \-cpio
and
.I \-ncpio
operators).
When producing a filename list for
.IR pax ,
find should always be invoked with
.I \-depth
since this makes it possible to extract write-protected directories
for users other than the super-user.
If
.I files
are given on the command line,
they are included in the archive
in the same manner as described above
and standard input is not read.
.PP
The
.B \-rw
options selects
.RI ` copy '
mode;
a list of
.I files
is read from standard input
or taken from the command line
as described for
.IR \-w ;
files are copied to the specified
.IR directory ,
preserving attributes as described for
.IR \-r .
Special files are re-created in the target hierarchy,
and hard links between copied files are preserved.
.PP
When a premature end-of-file is detected with
.I \-r
and
.I \-w
and the archive is a block or character special file,
the user is prompted for new media.
.PP
The following options alter the behavior of
.IR pax :
.TP
.B \-a
Append files to the archive.
The archive must be seekable,
such as a regular file or a block device,
or a tape device capable of writing between filemarks.
.TP
\fB\-b\fI size\fR
Blocks input and output archives at
.I size
byte records.
.TP
.B \-c
Reverses the sense of patterns
such that a file that does not match any of the patterns
is selected.
.TP
.B \-d
Causes
.I pax
to ignore files below directories.
In read mode,
patterns matching directories
cause only the directory itself to extracted,
files below will be ignored
unless another pattern applies to them.
In write mode,
arguments or standard input lines referring to directories
do not cause files below the respective directory
to be archived.
.TP
\fB\-f\fI\ file\fR
Selects a
.I file
that is read with the
.I \-r
option instead of standard input
or written with the
.I \-w
option instead of standard output.
.TP
.B \-i
Rename files interactively.
Before a file is extracted from the archive,
its file name is printed on standard error
and the user is prompted to specify a substitute file name.
If the line read from the terminal is empty,
the file is skipped;
if the line consists of a single dot,
the name is retained;
otherwise,
the line forms the new file name.
.TP
.B \-k
Causes existing files not to be overwritten.
.TP
.B \-K
Try to continue operation on read errors and invalid headers.
If an archive contains another archive,
files from either archive may be chosen.
.TP
.B \-l
Link files instead of copying them with
.I \-rw
if possible.
.TP
.B \-L
Follow symbolic links when reading files with
.I \-w
or
.IR \-rw .
.TP
.B \-n
If any
.I pattern
arguments are present,
each pattern can match exactly one archive member;
further members that could match the particular pattern are ignored.
Without
.I pattern
arguments,
only the first occurence of
a file that occurs more than once in the archive
is selected, the following are ignored.
.TP
\fB-o\ \fIoption\fB,\fR[\fIoption\fB,\fR\|...]
Specifies options as described for \fI\-x pax\fR.
.TP
\fB\-p\ \fIstring\fR
Specifies which file modes are to be preserved or ignored.
.I string
may contain one or more of
.RS
.TP
.B a
Inhibits preservation of file access times.
.TP
.B e
Causes preservation of every possible mode, ownership and time.
.TP
.B m
Inhibits preservation of file modification times.
.TP
.B o
Causes preservation of owner and group IDs
(for the super-user only).
.TP
.B p
Causes preservation of file mode bits
regardless of the umask
(see
.IR umask (2)).
.RE
.TP
\fB\-s\ /\fIregular expression\fB/\fIreplacement\fB/\fR[\fBgp\fR]
Modifies file names in a manner similar to that described in
.IR ed (1).
The
.I p
flag causes each modified file name to printed.
Any character can be used as delimiter instead of
.RI ` / '.
If a file name is empty after the replacement is done,
the file is ignored.
.TP
.B \-t
Resets the access times of files
that were included in the archive with
.IR \-r .
.TP
.B \-u
In read mode,
.I pax
will not overwrite existing target files
that were modified more recently than the file in the archive
when this option is given.
In write mode,
.I pax
will read the archive first.
It will then only append those files to the archive
that are not already included
or were more recently modified.
.TP
.B \-v
Prints the file names of archived or extracted files with
.I \-r
and
.I \-w
and a verbose output format
if neither of them is given.
.TP
\fB\-x\fI header\fR
Specifies the archive header format to be one of:
.sp
.in +6
.TS
lfB l.
\fBnewc\fR	SVR4 ASCII cpio format;\ 
\fBcrc\fR	SVR4 ASCII cpio format with checksum;\ 
\fBsco\fR	T{
SCO UnixWare 7.1 ASCII cpio format;
T}
\fBscocrc\fR	T{
SCO UnixWare 7.1 ASCII cpio format with checksum;
T}
\fBodc\fR	T{
traditional ASCII cpio format, as standardized in IEEE Std. 1003.1, 1996;
T}
\fBcpio\fR	T{
same as \fIodc\fR;
T}
\fBbin\fR	binary cpio format;
\fBbbs\fR	byte-swapped binary cpio format;
\fBsgi\fR	T{
SGI IRIX extended binary cpio format;
T}
\fBcray\fR	T{
Cray UNICOS 9 cpio format;
T}
\fBcray5\fR	T{
Cray UNICOS 5 cpio format;
T}
\fBdec\fR	T{
Digital UNIX extended cpio format;
T}
\fBtar\fR	tar format;
\fBotar\fR	old tar format;
\fBustar\fR	T{
IEEE Std. 1003.1, 1996 tar format;
T}
.T&
l s.
\fBpax\fR[\fB:\fIoption\fB,\fR[\fIoption\fB,\fR\|...]]
.T&
l l.
\&	T{
IEEE Std. 1003.1, 2001 pax format.
Format-specific \fIoptions\fR are:
.in +2n
.ti 0
.br
\fBlinkdata\fR
.br
For a regular file which has multiple hard links,
the file data is stored once for each link in the archive,
instead of being stored for the first entry only.
This option must be used with care
since many implementations are unable
to read the resulting archive.
.ti 0
.br                                                                            
\fBtimes\fR
.br
Causes the times of last access and last modification
of each archived file
to be stored in an extended \fIpax\fR header.
This in particular allows the time of last access
to be restored when the archive is read.
.br
.in -2n
T}
\fBsun\fR	T{
Sun Solaris 7 extended tar format;
T}
\fBbar\fR	T{
SunOS 4 bar format;
T}
\fBgnu\fR	T{
GNU tar format;
T}
\fBzip\fR[\fB:\fIcc\fR]	T{
zip format with optional compression method.
If \fIcc\fR is one of
\fBen\fR (normal, default),
\fBex\fR (extra),
\fBef\fR (fast),
or
\fBes\fR (super fast),
the standard \fIdeflate\fR compression is used.
\fBe0\fR selects no compression,
and
\fBbz2\fR selects \fIbzip2\fR compression.
T}
.TE
.in -6
.sp
This option is ignored with
.I \-r
unless the
.I \-K
option is also present.
The default for
.I \-w
is traditional ASCII cpio
.I (odc)
format.
.PP
.ne 30
Characteristics of archive formats are as follows:
.sp
.TS
allbox;
l r r r l
lfB1 r2 n2 r2 c.
	T{
.ad l
maximum user/\%group id
T}	T{
.ad l
maximum file size
T}	T{
.ad l
maximum pathname length
T}	T{
.ad l
bits in dev_t (major/minor)
T}
\-x\ bin	65535	2 GB\ 	256	\ 16
\-x\ sgi	65535	9 EB\ 	256	\ 14/18
T{
\-x\ odc
T}	262143	8 GB\ 	256	\ 18
\-x\ dec	262143	8 GB\ 	256	\ 24/24
T{
\-x\ newc,
\-x\ crc
T}	4.3e9	4 GB\ 	1024	\ 32/32
T{
\-x\ sco, \-x\ scocrc
T}	4.3e9	9 EB\ 	1024	\ 32/32
T{
\-x\ cray, \-x\ cray5
T}	1.8e19	9 EB\ 	65535	\ 64
\-x\ otar	2097151	8 GB\ 	99	\ n/a
T{
\-x\ tar,
\-x\ ustar
T}	2097151	8 GB\ 	256 (99)	\ 21/21
\-x\ pax	1.8e19	9 EB\ 	65535	\ 21/21
\-x\ sun	1.8e19	9 EB\ 	65535	\ 63/63
\-x\ gnu	1.8e19	9 EB\ 	65535	\ 63/63
\-x\ bar	2097151	8 GB\ 	427	\ 21
\-x\ zip	4.3e9	9 EB\ 	60000	\ 32
.TE
.sp
.PP
The byte order of
.B binary
cpio archives
depends on the machine
on which the archive is created.
Unlike some other implementations,
.I pax
fully supports
archives of either byte order.
.I \-x\ bbs
can be used to create an archive
with the byte order opposed to that of the current machine.
.PP
The
.B sgi
format extends the binary format

ex-vi/Changes
Release ...
* The screen buffers for visual mode are now dynamically allocated, so
  vi usually does not return to ex mode with "screen too large" when the
  terminal is resized on a large monitor anymore.
* ex can now edit files with lines of arbitrary length. vi currently only
  shows the beginning of a line that does not fit onto the screen in its
  entirety.
* Viewing executables and compressed files is no longer inhibited.

Release 3/25/05
* vi no longer dies with a segmentation fault if a line does not fit on the
  screen after an insertion.
* The 'p' command now works correctly if the buffer contains a partial line
  with multibyte characters.
* Traditional regular expressions sometimes failed to operate correctly
  since the last release.

Release 2/25/05
* Traditional regular expressions can now be used with multibyte characters.
* When the 'ignorecase' option is toggled, saved regular expressions are now
  updated accordingly. (P)
* If a line began with a tabulator and another tabulator was inserted with
  the cursor located on the first tabulator, the display was not updated
  appropriately since the last release (Bugreport by Matthew Fischer). (P)
* Fixed a segmentation fault that occured in multibyte locales when operation
  was continued after vi had been stopped by ^Z, with the cursor positioned
  in a line longer than the terminal width as the last line of the current
  screen.
* Made multicolumn characters work in lines longer than the screen width in
  visual mode (Bugreport by Matthew Fischer).
* Made it work for Big5 locales (Patches by Matthew Fischer).
* Fixed a problem with the 'r' command in EUC-JP and Big5 locales (Bugreport
  by Matthew Fischer).
* The insertion of multicolumn characters and tab characters in multibyte
  locales now works with terminals that have the 'ic' but no 'im' termcap
  capability (Bugreport by Matthew Fischer).
* The argument to the -w option is correctly recognized now.
* If the SHELL environment variable is set to the empty string, it is now
  ignored.
* A non-null exit status is now returned if a file could not be opened, if
  an invalid address is given for a command, or if a tag is not found.
* If the match for a substitution is of length zero, a line overflow is
  now avoided when compiled with the 'UNIX(R) Regular Expression Library'.
* When moving left while the cursor is positioned over a multicolumn
  character at the end of the line, the bell is rung now (Bugreport by
  Matthew Fischer).
* When moving up or down to a row with different column arrangement while
  the cursor is positioned over a multicolumn character, the leftmost
  character above the original position is chosen in the new row.
* If both the -t and the -c option are given, the -t option is now
  processed first, i.e. the command is executed at the position where
  the tag was found.
* The -w option now also sets the scroll size for the 'z' command.
* When the name of a nonexisting file is given with the 'edit' ex command,
  the command now succeeds and prints a "[New file]" message.
* If standard output is not a terminal, no '\r' is written at the end of
  printed lines anymore.
* The 'source' ex command now works if command input comes from a pipe or
  regular file.
* Ex does not exit on errors immediately anymore if standard input is not
  a terminal but a pipe or regular file.
* The 'substitute' ex command can now be abbreviated as 'sub', 'subst' etc.
* A new version of mapmalloc.c that is derived from Unix 7th Edition code
  has been introduced.
* If the 'next!' ex command is given and the autowrite option is set, the
  current file is not written anymore.

Release 1/19/05
* The last release erroneously made 'X' work like 'x' in visual mode. It now
  deletes the character before the cursor again as documented (Bugreport by
  Matthew Fischer). (P)
* When a multicolumn character was replaced by another multicolumn character
  in insert mode, the display was not updated appropriately with terminals
  other than xterm.
* Various rendering errors happened with multicolumn characters if they
  started at an even column (counting from 1 upwards).
* When a multicolumn character was inserted and then replaced, the visual
  screen representation was sometimes not updated accordingly.
* Undoing the replacement of a multicolumn character by a singlecolumn
  character twice made the singlecolumn character invisible.
* The 'cw' visual command with a multibyte character as last character of
  the affected word located at the end of the line left garbage bytes past
  the end of the new text.
* Visual 'U' followed by 'u' lead to garbage on the screen when multibyte
  characters were present on the changed line.
* The position of '$' when changing text was fixed for cases where the first
  changed character had multiple columns but the last one had not.
* The handling of multicolumn characters was fixed for terminals without the
  IM (insert mode) capability. It is unlikely that such terminals actually
  exist, but vi will use the corresponding code for open mode if a termcap
  entry is not available.
* When an illegal multibyte sequence is entered in vi insert mode, no garbage
  bytes are generated anymore when the insert mode is left.
* The '\u', '\l', '\U', and '\L' substitution sequences work with multibyte
  characters now.
* Handle character case conversions with the '~' vi command correctly if the
  length of the converted multibyte sequence is smaller than the original one.

heirloom/cpio/pax.c
/*
 * cpio - copy file archives in and out
 *
 * Gunnar Ritter, Freiburg i. Br., Germany, April 2003.
 */
/*
 * Copyright (c) 2003 Gunnar Ritter
 *
 * This software is provided 'as-is', without any express or implied
 * warranty. In no event will the authors be held liable for any damages
 * arising from the use of this software.
 *
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute
 * it freely, subject to the following restrictions:
 *
 * 1. The origin of this software must not be misrepresented; you must not
 *    claim that you wrote the original software. If you use this software
 *    in a product, an acknowledgment in the product documentation would be
 *    appreciated but is not required.
 *
 * 2. Altered source versions must be plainly marked as such, and must not be
 *    misrepresented as being the original software.
 *
 * 3. This notice may not be removed or altered from any source distribution.
 */

#if __GNUC__ >= 3 && __GNUC_MINOR__ >= 4
#define	USED	__attribute__ ((used))
#elif defined __GNUC__
#define	USED	__attribute__ ((unused))
#else
#define	USED
#endif
static const char sccsid[] USED = "@(#)pax.sl	1.16 (gritter) 6/1/04";

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdio.h>
#include <stdarg.h>
#include <stdlib.h>
#include <string.h>
#include <fnmatch.h>
#include <dirent.h>
#include <regex.h>
#include <wchar.h>
#include <time.h>
#include <inttypes.h>

#include "iblok.h"
#include "cpio.h"

static char		**files;
static int		filec;
static struct iblok	*filinp;
static char		*path;
static size_t		pathsz;

static void		setpres(const char *);
static size_t		ofiles_pax(char **, size_t *);
static void		prtime_pax(time_t);
static void		parsesub(char *);

void
flags(int ac, char **av)
{
	const char	optstring[] = "rwab:cdf:ikKlLno:p:s:tuvx:X";
	int	i;
	int	illegal = 0;

	pax = 1;
	dflag = 1;
	uflag = 1;
	ofiles = ofiles_pax;
	prtime = prtime_pax;
	while ((i = getopt(ac, av, optstring)) != EOF) {
		switch (i) {
		case 'r':
			if (action && action != 'i')
				action = 'p';
			else
				action = 'i';
			break;
		case 'w':
			if (action && action != 'o')
				action = 'p';
			else
				action = 'o';
			break;
		case 'a':
			Aflag = 1;
			break;
		case 'b':
			if ((blksiz = atol(optarg)) <= 0)
				msg(4, -2,
					"Illegal size given for -b option.\n");
			Cflag = 1;
			break;
		case 'c':
			fflag = 1;
			break;
		case 'd':
			pax_dflag = 1;
			break;
		case 'f':
			Oflag = Iflag = optarg;
			break;
		case 'i':
			rflag = 1;
			break;
		case 'k':
			pax_kflag = 1;
			break;
		case 'K':
			kflag = 1;
			break;
		case 'l':
			lflag = 1;
			break;
		case 'L':
			Lflag = 1;
			break;
		case 'n':
			pax_nflag = 1;
			break;
		case 'o':
			pax_options(optarg, 1);
			break;
		case 'p':
			setpres(optarg);
			break;
		case 's':
			pax_sflag = 1;
			parsesub(optarg);
			break;
		case 't':
			aflag = 1;
			break;
		case 'u':
			uflag = 0;
			pax_uflag = 1;
			break;
		case 'v':
			vflag = 1;
			break;
		case 'x':
			if (strcmp(optarg, "cpio") == 0)
				fmttype = FMT_ODC;
			else {
				if (setfmt(optarg) < 0)
					illegal = 1;
			}
			break;
		case 'X':
			pax_Xflag = 1;
			break;
		default:
			illegal = 1;
		}
	}
	switch (action) {
	case 0:
		if (rflag || pax_kflag || pax_uflag || pax_preserve)
			illegal = 1;
		action = 'i';
		tflag = 1;
		setvbuf(stdout, NULL, _IOLBF, 0);
		/*FALLTHRU*/
	case 'i':
		if (aflag || pax_Xflag || lflag)
			illegal = 1;
		for (i = optind; i < ac; i++) {
			addg(av[i], 0);
			if (pax_dflag == 0) {
				char	*da;
				int	j;

				da = smalloc(strlen(av[i]) + 2);
				for (j = 0; av[i][j]; j++)
					da[j] = av[i][j];
				da[j++] = '/';
				da[j++] = '*';
				da[j] = 0;
				addg(da, 1);
				free(da);
			}
		}
		break;
	case 'o':
		if (fflag || pax_kflag || pax_nflag || kflag)
			illegal = 1;
		if (Aflag && Oflag == NULL) {
			msg(3, 0, "-a requires the -f option\n");
			illegal = 1;
		}
		if (optind != ac) {
			files = &av[optind];
			filec = ac - optind;
		} else
			filinp = ib_alloc(0, 0);
		if (pax_uflag)
			Aflag = 1;
		if (Aflag == 0 && fmttype == FMT_NONE)
			fmttype = FMT_ODC;
		break;
	case 'p':
		if (fflag || blksiz || Oflag || Iflag || fmttype != FMT_NONE ||
				kflag)
			illegal = 1;
		if (optind == ac)
			illegal = 1;
		else if (optind+1 != ac) {
			files = &av[optind];
			filec = ac - optind - 1;
			optind = ac - 1;
		} else
			filinp = ib_alloc(0, 0);
		break;
	}
	if (illegal)
		usage();
}

void
usage(void)
{
	fprintf(stderr, "USAGE:\n\
\t%s [-cdnvK] [-b size] [-f file] [-s replstr] [-x hdr] [patterns]\n\
\t%s -r[cdiknuvK] [-b size] [-f file] "/*[-o options] */"[-p priv] [-s replstr] [-x hdr] [patterns]\n\
\t%s -w[adituvLX] [-b size] [-f file] [-o options] [-s replstr] [-x hdr] [files]\n\
\t%s -rw[diklntuvLX] [-p priv] [-s replstr] [files] directory\n",
		progname, progname, progname, progname);
	exit(1);
}

static void
setpres(const char *s)
{
	s--;
	while (*++s) {
		pax_preserve &= ~PAX_P_EVERY;
		switch (*s) {
		case 'a':
			pax_preserve |= PAX_P_ATIME;
			break;
		case 'e':
			pax_preserve |= PAX_P_EVERY;
			break;
		case 'm':
			pax_preserve |= PAX_P_MTIME;
			break;
		case 'o':
			pax_preserve |= PAX_P_OWNER;
			break;
		case 'p':
			pax_preserve |= PAX_P_MODE;
			break;
		default:
			msg(2, 0, "ignoring unknown option \"-p%c\"\n",
					*s&0377);
		}
	}
	if (pax_preserve & PAX_P_EVERY)
		pax_preserve |= PAX_P_OWNER|PAX_P_MODE;
}

int
gmatch(const char *s, const char *p)
{
	int	val;
#ifdef	__GLIBC__
	/* avoid glibc's broken [^...] */
	extern char	**environ;
	char	**savenv = environ;
	char	*newenv[] = { "POSIXLY_CORRECT=", NULL };
	environ = newenv;
#endif	/* __GLIBC__ */
	val = fnmatch(p, s, 0) == 0;
#ifdef	__GLIBC__
	environ = savenv;
#endif	/* __GLIBC__ */
	return val;
}

static const char *
nextfile(void)
{
	char	*line = NULL;
	size_t	linsiz = 0, linlen;

	if (filinp) {
		if ((linlen=ib_getlin(filinp, &line, &linsiz, srealloc)) == 0) {
			filinp = NULL;
			return NULL;
		}
		if (line[linlen-1] == '\n')
			line[--linlen] = '\0';
		return line;
	} else if (filec > 0) {
		filec--;
		return *files++;
	} else
		return NULL;
}

static size_t
catpath(size_t pend, const char *base)
{
	size_t	blen = strlen(base);

	if (pend + blen + 2 >= pathsz)
		path = srealloc(path, pathsz = pend + blen + 16);
	if (pend == 0 || path[pend-1] != '/')
		path[pend++] = '/';
	strcpy(&path[pend], base);
	return pend + blen;
}

/*
 * Descend the directory hierarchies given using stdin or arguments
 * and return file names one per one.
 */
static size_t
ofiles_pax(char **name, size_t *namsiz)
{
	static DIR	**dt;
	static int	dti, dts;
	static int	*pend;
	static dev_t	*curdev;
	static ino_t	*curino;
	struct stat	st;
	struct dirent	*dp;
	const char	*nf;
	int	i;

	if (dt == NULL) {
		dt = scalloc(dts = 1, sizeof *dt);
		pend = scalloc(dts, sizeof *pend);
		curdev = scalloc(dts, sizeof *curdev);
		curino = scalloc(dts, sizeof *curino);
	}
	for (;;) {
		if (dti >= 0 && dt[dti] != NULL) {
			if ((dp = readdir(dt[dti])) != NULL) {
				if (dp->d_name[0] == '.' &&
						(dp->d_name[1] == '\0' ||
						 dp->d_name[1] == '.' &&
						 dp->d_name[2] == '\0'))
					continue;
				if (dti+1 <= dts) {
					dt = srealloc(dt, sizeof *dt * ++dts);
					pend = srealloc(pend, sizeof *pend*dts);
					curdev = srealloc(curdev, sizeof *curdev
							* dts);
					curino = srealloc(curino, sizeof *curino
							* dts);
				}
				pend[dti+1] = catpath(pend[dti], dp->d_name);
				if ((Lflag ? stat : lstat)(path, &st) < 0) {
					emsg(2, "Error with %s of \"%s\"",
							lflag? "stat" : "lstat",
							path);
					errcnt++;
				} else if ((st.st_mode&S_IFMT) == S_IFDIR &&
						(pax_Xflag == 0 ||
						 curdev[0] == st.st_dev)) {
					if (Lflag) {
						for (i = 0; i <= dti; i++)
							if (st.st_dev ==
								curdev[i] &&
								st.st_ino ==
								curino[i])
								break;
						if (i <= dti)
							break;
					}
					if ((dt[dti+1]=opendir(path)) == NULL) {
						emsg(2, "Cannot open directory "
								"\"%s\"", path);
						errcnt++;
					} else {
						dti++;
						curdev[dti] = st.st_dev;
						curino[dti] = st.st_ino;
						continue;
					}
				} else
					break;
			} else {
				path[pend[dti]] = '\0';
				closedir(dt[dti]);
				dt[dti--] = NULL;
				break;
			}
		} else {
			while ((nf = nextfile()) != NULL && stat(nf, &st) < 0) {
				emsg(2, "Error with stat of \"%s\"", nf);
				errcnt++;
			}
			if (nf == NULL)
				return 0;
			dti = 0;
			if (path)
				free(path);
			pend[dti] = strlen(nf);
			strcpy(path = smalloc(pathsz = pend[dti]+1), nf);
			if (pax_dflag || (st.st_mode&S_IFMT) != S_IFDIR) {
				dti = -1;
				break;
			}
			curdev[dti] = st.st_dev;
			curino[dti] = st.st_ino;
			if ((dt[dti] = opendir(path)) == NULL) {
				emsg(2, "Cannot open directory \"%s\"", path);
				errcnt++;
			}
		}
	}
	if (*name == NULL || *namsiz < pathsz) {
		free(*name);
		*name = smalloc(*namsiz=pathsz);
	}
	strcpy(*name, path);
	return pend[dti+1];
}

struct	pax_had {
	struct pax_had	*p_next;
	const char	*p_name;
	time_t		p_mtime;
};

static int	pprime = 7919;

static int
phash(const char *s)
{
	uint32_t	h = 0, g;

	s--;
	while (*++s) {
		h = (h << 4) + (*s & 0377);
		if (g = h & 0xf0000000) {
			h = h ^ (g >> 24);
			h = h ^ g;
		}
	}
	return h % pprime;
}

static int
plook(const char *name, struct pax_had **pp)
{
	static struct pax_had	**pt;
	uint32_t	h, had;

	if (pt == NULL)
		pt = scalloc(pprime, sizeof *pt);
	(*pp) = pt[h = phash(name)];
	while (*pp != NULL) {
		if (strcmp((*pp)->p_name, name) == 0)
			break;
		*pp = (*pp)->p_next;
	}
	had = *pp != NULL;
	if (*pp == NULL) {
		*pp = scalloc(1, sizeof **pp);
		(*pp)->p_name = sstrdup(name);
		(*pp)->p_next = pt[h];
		pt[h] = *pp;
	}
	return had;
}

int
pax_track(const char *name, time_t mtime)
{
	struct pax_had	*pp;

	if ((pax_uflag == 0 || action != 'o') && (pax_nflag == 0 || patterns))
		return 1;
	if (plook(name, &pp) != 0) {
		if (pax_uflag == 0)
			return 0;
		if (mtime > pp->p_mtime) {
			pp->p_mtime = mtime;
			return 1;
		}
		return 0;
	} else
		pp->p_mtime = mtime;
	return 1;
}

static void
prtime_pax(time_t t)
{
	char	b[30];
	time_t	now;

	time(&now);
	if (t > now || t < now - (6*30*86400))
		strftime(b, sizeof b, "%b %e  %Y", localtime(&t));
	else
		strftime(b, sizeof b, "%b %e %H:%M", localtime(&t));
	printf(" %s ", b);
}

struct replacement {
	regex_t			r_re;
	const char		*r_rhs;
	int			r_nbra;
	enum {
		REPL_0	= 0,
		REPL_G	= 1,
		REPL_P	= 2
	}			r_flags;
} *rep;

#define	NBRA	9
static int	ren, res;
static int	mb_cur_max;

static wchar_t
nextc(char **sc, int *np)
{
	char	*p = *sc;
	wchar_t	wcbuf;
	int	len;

	if (**sc == '\0') {
		*np = 0;
		return 0;
	}
	if (mb_cur_max == 1 || (**sc&0200) == 0) {
		*np = 1;
		return *(*sc)++ & 0377;
	}
	if ((len = mbtowc(&wcbuf, p, mb_cur_max)) < 0)
		msg(3, -2, "Invalid multibyte character for \"-s\" option\n");
	*np = len;
	*sc += len;
	return wcbuf;
}

static void
parsesub(char *s)
{
	int	len;
	char	*ps = NULL;
	wchar_t	seof = nextc(&s, &len);
	wint_t	c, d;
	int	nbra = 0;

	if (seof == 0)
		goto unt;
	mb_cur_max = MB_CUR_MAX;
	ps = s;
	do {
		if ((c = nextc(&s, &len)) == seof)
			break;
		if (c == '\\') {
			if ((c = nextc(&s, &len)) == '(')
				nbra++;
			continue;
		} else if (c == '[') {
			d = WEOF;
			do {
				if ((c = nextc(&s, &len)) == '\0')
					continue;
				if (d == '[' && (c == ':' || c == '.' ||
							c == '=')) {
					d = c;
					do {
						if ((c=nextc(&s, &len)) == '\0')
							continue;
					} while (c != d || *s != ']');
					nextc(&s, &len);
					c = WEOF; /* reset d and continue */
				}
				d = c;
			} while (c != ']');
		}
	} while (*s != '\0');
	if (c != seof)
	unt:	msg(3, -2, "Unterminated argument for \"-s\" option.\n");
	s[-len] = '\0';
	if (ren <= res)
		rep = srealloc(rep, ++res * sizeof *rep);
	if (regcomp(&rep[ren].r_re, ps, REG_ANGLES|REG_BADRANGE) != 0)
		msg(3, -2, "Regular expression error in \"-s\" option\n");
	rep[ren].r_rhs = s;
	rep[ren].r_nbra = nbra;
	while ((c = nextc(&s, &len)) != 0) {
		if (c == '\\')
			c = nextc(&s, &len);
		else if (c == seof)
			break;
	}
	rep[ren].r_flags = 0;
	if (c == seof) {
		s[-len] = '\0';
		while ((c = nextc(&s, &len)) != '\0') {
			switch (c) {
			case 'g':
				rep[ren].r_flags |= REPL_G;
				break;
			case 'p':
				rep[ren].r_flags |= REPL_P;
				break;
			default:
				msg(2, 0, "Ignoring unknown -s flag \"%c\"\n",
						c);
			}
		}
	}
	ren++;
}

#define	put(c)	((new = innew+1>=newsize ? srealloc(new, newsize+=32) : new), \
			new[innew++] = (c))

int
pax_sname(char **oldp, size_t *olds)
{
	char	*new = NULL;
	size_t	newsize = 0;
	regmatch_t	bralist[NBRA+1];
	int	c, i, k, l, y, z;
	int	innew = 0, ef = 0;
	char	*inp = *oldp;

	for (z = 0; z < ren; z++) {
	in:	if (regexec(&rep[z].r_re, inp, NBRA+1, bralist, ef) != 0) {
			if (ef == 0)
				continue;
			goto out;
		}
		for (i = 0; i < bralist[0].rm_so; i++)
			put(inp[i]);
		k = 0;
		while (c = rep[z].r_rhs[k++] & 0377) {
			y = -1;
			if (c == '&')
				y = 0;
			else if (c == '\\') {
				c = rep[z].r_rhs[k++] & 0377;
				if (c >= '1' && c < rep[z].r_nbra+'1')
					y = c - '0';
			}
			if (y >= 0)
				for (l = bralist[y].rm_so; l < bralist[y].rm_eo;
						l++)
					put(inp[i]);
			else
				put(c);
		}
		k = innew;
		for (i = bralist[0].rm_eo; inp[i]; i++)
			put(inp[i]);
		put('\0');
		if (rep[z].r_flags & REPL_G) {
			ef = REG_NOTBOL;
			inp = &inp[bralist[0].rm_eo];
			innew = k;
			if (bralist[0].rm_so == bralist[0].rm_eo) {
				if (inp[0] && (nextc(&inp, &l), inp[0]))
					innew++;
				else
					goto out;
			}
			goto in;
		}
	out:	if (rep[z].r_flags & REPL_P)
			fprintf(stderr, "%s >> %s\n", *oldp, new);
		free(*oldp);
		*oldp = new;
		*olds = newsize;
		return *new != '\0';
	}
	return 1;
}

void
pax_onexit(void)
{
	struct glist	*gp;

	for (gp = patterns; gp; gp = gp->g_nxt) {
		if (gp->g_art)
			continue;
		if (gp->g_gotcha == 0 && (gp->g_nxt == NULL ||
					gp->g_nxt->g_art == 0 ||
					gp->g_gotcha == 0)) {
			msg(3, 0, "Pattern not matched: \"%s\"\n", gp->g_pat);
			errcnt++;
		}
	}
}
